pkg/autoscaler/scaling/autoscaler.go

```diff
@@ -245,8 +244,49 @@ func (a *Autoscaler) Scale(ctx context.Context, now time.Time) (desiredPodCount
 	return desiredPodCount, int32(excessBCF), true
 }
 
+func clampInclusiveRange(valueToClamp float64, lowerBound float64, upperBound float64) float64 {
+	return math.Min(math.Max(valueToClamp, lowerBound), upperBound)
+}
+
+// Query functions
+func (a *Autoscaler) currentlyStable() bool {
+	return a.panicTime.IsZero()
+}
+
+func (a *Autoscaler) currentlyPanicking() bool {
+	return !a.currentlyStable()
+}
+
+func (a *Autoscaler) stableWindowHasElapsed(spec *DeciderSpec, now time.Time) bool {
+	return a.panicTime.Add(spec.StableWindow).Before(now)
+}
+
 func (a *Autoscaler) currentSpecAndPC() (*DeciderSpec, resources.EndpointsCounter) {
 	a.specMux.RLock()
 	defer a.specMux.RUnlock()
 	return a.deciderSpec, a.podCounter
 }
+
+// Panic functions
+func (a *Autoscaler) updatePanic(panicTime time.Time, maxPanicPods int32) {
+	a.panicTime = panicTime
+	a.maxPanicPods = maxPanicPods
+}
+
+func (a *Autoscaler) startPanicking(now time.Time, logger *zap.SugaredLogger) {
+	// Begin panicking when we cross the threshold in the panic window.
```

Comment:
The comment is out of place here.
Also remove the empty lines here and below. The code feel unnaturally spaced.

Reply:
Updated in 94457ef.